src/training/stylegan_adversarial.py

```diff
@@ -0,0 +1,261 @@
+                 scheduler_G: optim.lr_scheduler._LRScheduler,
+                 scheduler_D: optim.lr_scheduler._LRScheduler,
+                 train_dataloader: DataLoader,
+                 val_dataloader: DataLoader,
+                 storage: Storage,
+                 logger: Logger,
+                 total_epochs: int,
+                 device: str,
+                 ocr_coef: float,
+                 cycle_coef: float,
+                 recon_coef: float,
+                 emb_coef: float,
+                 perc_coef: float,
+                 tex_coef: float,
+                 adv_coef: float,
+                 ocr_loss: nn.Module,
+                 typeface_loss: nn.Module,
+                 perc_loss: nn.Module,
+                 cons_loss: nn.Module,
+                 adv_loss: nn.Module
+                 ):
+
+        self.device = device
+        self.model_G = model_G
+        self.model_D = model_D
+        self.optimizer_G = optimizer_G
+        self.optimizer_D = optimizer_D
+        self.scheduler_G = scheduler_G
+        self.scheduler_D = scheduler_D
+        self.train_dataloader = train_dataloader
+        self.val_dataloader = val_dataloader
+        self.total_epochs = total_epochs
+        self.logger = logger
+        self.storage = storage
+        self.ocr_coef = ocr_coef
+        self.cycle_coef = cycle_coef
+        self.recon_coef = recon_coef
+        self.emb_coef = emb_coef
+        self.perc_coef = perc_coef
+        self.tex_coef = tex_coef
+        self.adv_coef = adv_coef
+        self.typeface_loss = typeface_loss.to(device)
+        self.ocr_loss = ocr_loss.to(device)
+        self.perc_loss = perc_loss.to(device)
+        self.cons_loss = cons_loss.to(device)
+        self.adv_loss = adv_loss.to(device)
+        self.style_embedder = style_embedder
+        self.content_embedder = content_embedder
+    
+    def set_requires_grad(self, net: nn.Module, requires_grad: bool = False):
+        if net is not None:
+            for param in net.parameters():
+                param.requires_grad = requires_grad
+    
+    def model_D_adv_loss(self, 
+                         style_imgs: torch.Tensor, 
+                         content_embeds: torch.Tensor, 
+                         style_embeds: torch.Tensor
+                         ):
+        pred_D_fake = self.model_D(self.model_G(content_embeds, style_embeds).detach())
+        fake = torch.tensor(0.).expand_as(pred_D_fake).to(self.device)
+        loss_D_fake = self.adv_loss(pred_D_fake, fake)
+
+        pred_D_real = self.model_D(style_imgs)
+        valid = torch.tensor(1.).expand_as(pred_D_real).to(self.device)
+        loss_D_real = self.adv_loss(pred_D_real, valid) 
+
+        return (loss_D_real + loss_D_fake) * 0.5
+
+    def model_G_adv_loss(self, res: torch.Tensor):
+        pred_D_fake = self.model_D(res)
+        valid = torch.tensor(1.).expand_as(pred_D_fake).to(self.device)
+
+        return self.adv_loss(pred_D_fake, valid)
```

Comment:
These could be a separate loss object maybe

Reply:
Agreed

Reply:
not sure because we need params of the trainer in these funcs

Reply:
True, I think it’s okay both ways then, it’s up to you

Reply:
We could pass them along the batches, or when constructing the loss. But yeah, it's not a big deal.